### ISSUE TITLE
Training difficulties

### ISSUE THREAD

**Issue by LouisSerrano** [lucidrains/magvit2-pytorch#4]
Hi, I am experiencing some difficulties during the training of magvit2. I don't know if I made some mistakes somewhere or where the problem might be coming from.

It seems that my understanding of the paper might me be erroneous, I tried with 2 codebooks of size 512 and I can't seem to fit the training data. The training is really unstable. I tried to replace the LFQ with a classical VQ and it was more stable and was able to converge.
What is the config that you tried for training the model ?

**Comment by lucidrains** [comment]
@LouisSerrano hey Louis and thank you for reporting this

that is disappointing to hear, as I had high hopes for LFQ. there is the possibility that I implemented it incorrectly, but the authors had already given it a code review, so that seems less likely. i'll be training the model tonight (regaining access to my deep learning rig) so i can see this instability for myself.

**Comment by lucidrains** [comment]
@LouisSerrano as a silver lining, i guess your experiments show the rest of the framework to be working ok. i can run a few toy experiments on LFQ this morning before running it on a big image dataset tonight and see what the issue is

**Comment by lucidrains** [comment]
@LouisSerrano just to make sure, are you on the latest version of the `vector-quantize-pytorch` library?

**Comment by LouisSerrano** [comment]
Ok thank you very much, it might simply be an error from my side, probably in the configuration. I used the model config that you suggested in the Readme.md, and for LFQ I used 1 codebook with size 512.

**Comment by LouisSerrano** [comment]
@lucidrains yes I am using 1.10.4

**Comment by lucidrains** [comment]
@LouisSerrano no that should be fine, you could try increasing to 2048 or 4096, but shouldn't make a big difference

**Comment by lucidrains** [comment]
@LouisSerrano this is kind of a stab in the dark, but would you like to try adding `lfq_activation = nn.Tanh(), # from torch import nn` to your `VideoTokenizer` init?

**Comment by lucidrains** [comment]
they didn't have this in the paper, but i think it should make sense

**Comment by lucidrains** [comment]
@LouisSerrano let me do some toy experiments right now and make sure it isn't some obvious bug

**Comment by lucidrains** [comment]
@LouisSerrano thank you! 🙏

**Comment by LouisSerrano** [comment]
@lucidrains  Thanks for the awesome work !

**Comment by lucidrains** [comment]
@LouisSerrano yea no problem! well, we'll see if this work pans out. thank you for attempting to replicate!

**Comment by lucidrains** [comment]
<img width="516" alt="Screen Shot 2023-11-02 at 7 25 07 AM" src="https://github.com/lucidrains/magvit2-pytorch/assets/108653/d94b4f8b-6a2e-46d3-9e8b-6a05287643e6">

**Comment by lucidrains** [comment]
for the toy task, LFQ looks ok compared to VQ. Tanh won't work, but you can try `Tanh x 10`

```python
lfq_activation = lambda x: torch.nn.functional.tanh(x) * 10,
```

**Comment by LouisSerrano** [comment]
Ok thanks I will try this ! I'll let you know if I encounter some issues. Also what kind of weights do you use for the commitment and entropy loss ?

**Comment by lucidrains** [comment]
@LouisSerrano i think for commitment loss it is ok to keep it at the value as regular VQ of `1.`, but i'm not sure about the respective per-sample and batch entropy

**Comment by lucidrains** [comment]
@LouisSerrano i just increased the batch entropy weight a bit, to what works for the toy task (just fashion mnist)

**Comment by LouisSerrano** [comment]
@lucidrains Ok great, thanks for the tips.

**Comment by lucidrains** [comment]
@LouisSerrano hey Louis, just noticed that you used the default layer structure from the readme. feel free to try the updated one and see how that fares with LFQ

would still be interested to know if `tanh(x) * 10` helps resolve your previous instability, if you had an experiment in progress.

**Comment by LouisSerrano** [comment]
This was with my previous config. I benchmarked against fsq. I also show the aux_loss, which is going crazy for lfq_tanh

<img width="684" alt="Capture d’écran 2023-11-02 à 16 51 12" src="https://github.com/lucidrains/magvit2-pytorch/assets/33746360/8dffa1df-b280-41ce-8da4-89b56f775994">

**Comment by lucidrains** [comment]
@LouisSerrano thank you! do you have the original unstable LFQ plot too? and wow, you actually gave FSQ a test drive; what level settings did you use for it?

**Comment by lucidrains** [comment]
@LouisSerrano did you compare it to baseline VQ by any chance?

**Comment by lucidrains** [comment]
that plot for FSQ looks very encouraging, maybe i'll add it tonight to the repository to further research

**Comment by LouisSerrano** [comment]
I think my main concern was on the auxiliary loss, I did not look to much into details but I assumed that somehow the model struggled to have a diverse codebook and good reconstructions
<img width="702" alt="Capture d’écran 2023-11-02 à 17 02 35" src="https://github.com/lucidrains/magvit2-pytorch/assets/33746360/0f9506ea-97d9-4ca2-b530-4eeca6789886">

**Comment by LouisSerrano** [comment]
@lucidrains I did not use VQ yet but I am gonna launch it tonight. So I'll let you know when I get the results. Also, I tested two recommended configurations for fsq: levels_big = [8,8,8,6,5] and levels_mid = [8,5,5,5]. On my simple problem I would have expected similar results for both and it is the case, so I am happy with it !

**Comment by LouisSerrano** [comment]
@lucidrains I will give a shot to your config, thanks again !

**Comment by lucidrains** [comment]
@LouisSerrano ok, let me know how VQ goes on your end! i'm hoping it is worse than both LFQ and FSQ, on that broken layer architecture you used (which is still a good benchmark)

if VQ outperforms both LFQ and FSQ on both the old and new layer architecture, then i don't know what to do

**Comment by lucidrains** [comment]
@LouisSerrano  i'll be running experiments tonight and making sure image pretraining is seamless

**Comment by LouisSerrano** [comment]
@lucidrains Don't worry, I will continue to inspect the code on my side. I made some adjustments to adapt the code for my dataset, so there still could be an issue somewhere only in my code. We are going to inspect the reconstructions on the same dataset with a different VQ implementation to see If I possibly made an error somewhere.

**Comment by lucidrains** [comment]
@LouisSerrano looking at your results, i'll put in some more work into FSQ today, make it break parity with LFQ. the magvit authors claim LFQ helped them break a SOTA while FSQ did not demonstrate that, but it is still early days.

**Comment by lucidrains** [comment]
@LouisSerrano added FSQ, hope it helps!

**Comment by LouisSerrano** [comment]
@lucidrains Thanks ! Yes I agree, it would be interesting to have the FSQ vs LFQ comparison for magvit2.

**Comment by lucidrains** [comment]
@LouisSerrano hey Louis, i had a big bug where i forgot to [include quantizer parameters](https://github.com/lucidrains/magvit2-pytorch/commit/9be329ab025901cc8e2db9e696bc1958d0884246)

may obsolete all your results, sorry!

**Comment by lucidrains** [comment]
@LouisSerrano i'm seeing lower reconstruction loss for LFQ

**Comment by LouisSerrano** [comment]
@lucidrains Thanks for this, that's good to hear ! On my data, which might be a bit different from the one in the paper, I get fsq > vq > lfq so far. But this does not mean that lfq does not work.
<img width="452" alt="Capture d’écran 2023-11-03 à 16 56 39" src="https://github.com/lucidrains/magvit2-pytorch/assets/33746360/aa69591d-465d-4a53-a71b-6d609f217b4c">

**Comment by lucidrains** [comment]
@LouisSerrano deleted my comment, because i realized i didn't make the number of codes equal lol. i'm not much of an experimentalist, as you can probably tell

is this with the updated repository with the projections into the codebook dimension being optimized? how many codes are you doing for all of them?

**Comment by lucidrains** [comment]
@LouisSerrano also, are the sampled reconstructions looking ok for your dataset?

**Comment by LouisSerrano** [comment]
@lucidrains I am using around 512 codes for each: [8, 6, 4, 3] for fsq, (512, 16) for vq and 2**9 for lfq.

**Comment by lucidrains** [comment]
thank you 🙏

**Comment by sekstini** [comment]
Interesting! I moved over to stable-audio-tools and [implemented FSQ there](https://github.com/Stability-AI/stable-audio-tools/commit/bd97c5714bce9efdf6984779a6a95011fe2663de), and it has been working surprisingly well for me so far. Thinking of trying out some ablations, including LFQ once I do proper training runs.

**Comment by lucidrains** [comment]
@sekstini awesome!

**Comment by sekstini** [comment]
![stacked_fsq](https://github.com/lucidrains/magvit2-pytorch/assets/127142660/93599510-1e2f-496f-a8f3-9f33c8312849)

Sneak peek of FSQ latents corresponding to 1 second of 48kHz audio ^^

**Comment by lucidrains** [comment]
@sekstini got [residual FSQ](https://github.com/lucidrains/vector-quantize-pytorch/blob/master/vector_quantize_pytorch/residual_fsq.py) done this morning

you are both getting me excited

**Comment by lucidrains** [comment]
[giving residual fsq a test drive with soundstream](https://wandb.ai/lucidrains/soundstream/reports/soundstream-with-residual-fsq--Vmlldzo1ODcyMjc1) will let this run throughout the day and check on it through the phone

**Comment by sekstini** [comment]
Doing some test runs comparing SFSQ (Stacked FSQ) with RVQ here:
https://wandb.ai/sekstini/smolgoose

Will maybe throw in LFQ later today.

**Comment by lucidrains** [comment]
hmm, not seeing much a difference, at least for soundstream. i'll keep trying though

**Comment by lucidrains** [comment]
have either of you seen any interesting papers dealing with loss balancing? (when there are like a gazillion auxiliary losses)

**Comment by RenShuhuai-Andy** [comment]
> for the toy task, LFQ looks ok compared to VQ. Tanh won't work, but you can try `Tanh x 10`
> 
> ```python
> lfq_activation = lambda x: torch.nn.functional.tanh(x) * 10,
> ```

Hi @lucidrains , I couldn't find any instances where the [lfq_activation](https://github.com/lucidrains/magvit2-pytorch/blob/main/magvit2_pytorch/magvit2_pytorch.py#L1068) is utilized. I wonder how this activation function works, and let me know if there's something I might be overlooking?

**Comment by JordanLazzaro** [comment]
@lucidrains @LouisSerrano when using FSQ in place of LFQ, are you producing a latent image/volume of size of (len(levels), H, W)? And if so, to get a token sequence for FSQ, are you just converting from codes to indices and flattening (ie. (len(levels), H, W) -> (1, H, W) -> (1*H*W))?

**Comment by denadai2** [comment]
> @sekstini got [residual FSQ](https://github.com/lucidrains/vector-quantize-pytorch/blob/master/vector_quantize_pytorch/residual_fsq.py) done this morning
> 
> you are both getting me excited

me too :D
